<?xml version="1.0" encoding="UTF-8"?>

<launch>
  <!-- MANDATORY PARAMETERS TO DEFINE -->
  <!-- Common Options:  -->
      <!--        /point_cloud_filter/lidar_depth_camera/point_cloud_filtered   -->
      <!--        /lidar/point_cloud  -->
      <!--        /hesai/pandar  -->
  <arg name="cloud_topic" default="/gt_box/livox/lidar_converted"/>
  
  <!-- The tracked frame of the External odometry. Transformation from this frame to tracked_sensor_Frame should exist. -->
  <!-- Reads as `pose of <frame> in external odometry frame` -->
  <!-- Common Options:  -->
      <!--   base   -->
      <!--   zed2i_base_link  -->
  <arg name="assumed_external_odometry_tracked_frame" default="livox"/>

  <!-- Currently only used for attitude initialization -->
  <!-- Common Options:  -->
  <!--   /imu_perception_kit   -->
  <!--   /sensors/imu  -->
  <arg name="imu_topic" default="/imu_perception_kit"/>

  <!-- Arguments -->
  <arg name="launch_prefix" default="" doc="gdb -ex run --args , valgrind --leak-check=full"/>
  <arg name="launch_rviz" default="true"/>

   <!-- These 3 topics are exclusive. Only one of them going to be used even if multiple exist. This is done to support multiple msg types. -->
  <arg name="odometry_topic" default="/dlio/lidar_map_odometry"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->
  <arg name="pose_stamped_topic" default="/point_cloud_filter/lidar_depth_camera/odometry_pose"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->
  <arg name="pose_stamped_with_covariance_topic" default="/state_estimator/pose_in_odom"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->
  <arg name="parameter_filename" default="param_velodyne_puck16.lua"/>
  <arg name="parameter_folder_path" default="$(find open3d_slam_ros)/param/"/>
  <arg name="map_saving_folder" default="$(find open3d_slam_ros)/data/maps/"/>
  <arg name="num_accumulated_range_data" default="1"/>

  <!-- Replay Options -->
  <arg name="is_read_from_rosbag" default="false"/>

  <!-- Main Node -->
  <node name="open3d_slam" pkg="open3d_slam_ros"
        type="mapping_node" output="screen"
        launch-prefix="$(arg launch_prefix)">
    <param name="cloud_topic" type="string" value="$(arg cloud_topic)"/>
    <param name="odometry_topic" type="string" value="$(arg odometry_topic)"/>
    <param name="pose_stamped_topic" type="string" value="$(arg pose_stamped_topic)"/>
    <param name="pose_stamped_with_covariance_topic" type="string" value="$(arg pose_stamped_with_covariance_topic)"/>
    <param name="imu_topic" type="string" value="$(arg imu_topic)"/>
    <param name="parameter_folder_path" type="string" value="$(arg parameter_folder_path)"/>
    <param name="assumed_external_odometry_tracked_frame" type="string" value="$(arg assumed_external_odometry_tracked_frame)"/>
    <param name="parameter_filename" type="string" value="$(arg parameter_filename)"/>
    <param name="num_accumulated_range_data" value="$(arg num_accumulated_range_data)"/>
    <param name="is_read_from_rosbag" value="$(arg is_read_from_rosbag)"/>
    <param name="map_saving_folder" value="$(arg map_saving_folder)"/>
  </node>

  <!-- Visualization -->
  <include
      file="$(find open3d_slam_ros)/launch/vis.launch"
      if="$(arg launch_rviz)">
  </include>

</launch>